### ISSUE TITLE
Early termination does not happen in corner case

### ISSUE THREAD

**Issue by ekcs** [open-policy-agent/opa#4894]
## Short description 
In OPA v0.42.1 (via Rego playground), early termination does not seem to take effect in a weird corner case. Or perhaps not a bug because early termination was never designed to work in this case.

Rules:
```rego
identify_func(x) := x

a {
    input["foo"] == 1
}

a {
    identify_func(2)    
}
```

Input:
```json
{
    "foo": 1
}
```

## Steps To Reproduce
1. Use the above rules with the above input. (Or go to [https://play.openpolicyagent.org/p/R8pdXCQqpQ](https://play.openpolicyagent.org/p/R8pdXCQqpQ))
3. Evaluate with coverage turned on
4. See that all the `a` rules are evaluated (lines 9 and 13 in the playground example)

## Expected behavior

Expect only one of the two rules to be evaluated due to early termination.

## Additional context

The behavior is very specific to a corner case. Line 17 and on give slight tweaks to the case which no longer produce the unexpected behavior.

```rego
# b
# doing the same logic but this time using a built-in
# function avoids the bug
b {
    input.foo == 1
}

b {
    is_number(2)
}

# c
# tweaking the first rule condition (still depends on input)
# also seems to avoid the bug

c {
    input["foo"]
}

c {
    identify_func(2)    
}
```

**Comment by srenatus** [comment]
Thanks for the report. It's curious, I'll have a closer look today.

**Comment by srenatus** [comment]
So, I've been digging in. It's been fun to dive into that part of the code base once again 😅 

What's going on is this: we've got some heuristic at play to determine when a rule (including non-built-in functions) is safe to early-exit from. It goes like this:

> All the rule's bodies' values are the same, and it's ground.

So for
```rego
identity_func(x) := x
```
we only have one rule, so the only value to care about is `x`. Is it ground? It is not. So EE is disabled for `identity_func`. Once a function in an eval path is not EE-enabled, EE can't happen. So this is what's hindering EE for the second rule body of `a` in your example. If we tried with `indentity_func(x) := 2`, for example, we'd find that only the second rule body of `a` is evaluated, because now, `identity_func` is EE-safe.

What this doesn't answer is what happens to `c`. I think for some reason, the result of the rule index lookup for `c` brings up the `input["foo"]` rule _first_, and we thus are able to early exit after it's been evaluated.

Maybe that's the key to improving this edge case. I'll dig in a little more.
